File: 3.x.x/crates-utopia/messaging/src/adapter/email/smtp.rs
//! PHP `Utopia\Messaging\Adapter\Email\SMTP`.

use std::net::{SocketAddr, TcpStream, ToSocketAddrs};
use std::time::Duration;

use lettre::address::Envelope;
use lettre::transport::smtp::authentication::{Credentials, Mechanism};
use lettre::transport::smtp::client::{Tls, TlsParameters};
use lettre::{Address, SmtpTransport, Transport};

use super::mime::Mime;
use super::TYPE;
use crate::adapter::{expect_email, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::Email;
use crate::php::php_empty_str;
use crate::response::{Response, ResponseData};

/// PHP `Utopia\SMTP\Encryption`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SMTPEncryption {
    /// PHP `Encryption::None`.
    None,
    /// PHP `Encryption::Implicit` (`ssl://`).
    Implicit,
    /// PHP `Encryption::StartTls` (`tls://`).
    StartTls,
    /// PHP `Encryption::Opportunistic` (`AutoTLS`).
    Opportunistic,
}

/// One parsed host entry: `(host, port, encryption)`.
pub type SMTPHost = (String, u16, SMTPEncryption);

/// PHP `Adapter\Email\SMTP`.
pub struct SMTP {
    base: AdapterBase,
    host: String,
    port: u16,
    username: String,
    password: String,
    smtp_secure: String,
    smtp_auto_tls: bool,
    x_mailer: String,
    timeout: u64,
    keep_alive: bool,
    timelimit: u64,
    client: parking_lot::Mutex<Option<SmtpTransport>>,
}

impl std::fmt::Debug for SMTP {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SMTP")
            .field("host", &self.host)
            .field("port", &self.port)
            .finish_non_exhaustive()
    }
}

impl SMTP {
    /// PHP `__construct`.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        host: impl Into<String>,
        port: u16,
        username: impl Into<String>,
        password: impl Into<String>,
        smtp_secure: impl Into<String>,
        smtp_auto_tls: bool,
        x_mailer: impl Into<String>,
        timeout: u64,
        keep_alive: bool,
        timelimit: u64,
    ) -> Result<Self, MessagingError> {
        let smtp_secure = smtp_secure.into();
        if !matches!(smtp_secure.as_str(), "" | "ssl" | "tls") {
            return Err(MessagingError::invalid_argument(
                "Invalid SMTP secure prefix. Must be \"\", \"ssl\" or \"tls\"",
            ));
        }
        Ok(Self {
            base: AdapterBase::default(),
            host: host.into(),
            port: if port == 0 { 25 } else { port },
            username: username.into(),
            password: password.into(),
            smtp_secure,
            smtp_auto_tls,
            x_mailer: x_mailer.into(),
            timeout: if timeout == 0 { 30 } else { timeout },
            keep_alive,
            timelimit: if timelimit == 0 { 30 } else { timelimit },
            client: parking_lot::Mutex::new(None),
        })
    }

    /// Convenience constructor with PHP defaults (`port=25`, empty credentials).
    pub fn with_host_port(host: impl Into<String>, port: u16) -> Result<Self, MessagingError> {
        Self::new(host, port, "", "", "", false, "", 30, false, 30)
    }

    /// PHP `disconnect()`.
    pub fn disconnect(&self) {
        *self.client.lock() = None;
    }

    /// PHP `hosts()` - public so unit tests can assert parsing without SMTP.
    #[must_use]
    pub fn hosts(&self) -> Vec<SMTPHost> {
        let mut hosts = Vec::new();
        for entry in self.host.split(';') {
            let entry = entry.trim();
            if entry.is_empty() {
                continue;
            }
            let mut encryption = self.encryption();
            let mut rest = entry.to_string();
            let lower = rest.to_ascii_lowercase();
            if let Some(stripped) = lower.strip_prefix("ssl://") {
                encryption = SMTPEncryption::Implicit;
                let prefix_len = entry.len() - stripped.len();
                rest = entry[prefix_len..].to_string();
            } else if let Some(stripped) = lower.strip_prefix("tls://") {
                encryption = SMTPEncryption::StartTls;
                let prefix_len = entry.len() - stripped.len();
                rest = entry[prefix_len..].to_string();
            }
            let (host, port) = parse_host_port(&rest, self.port);
            hosts.push((host, port, encryption));
        }
        if hosts.is_empty() {
            vec![(self.host.clone(), self.port, self.encryption())]
        } else {
            hosts
        }
    }

    fn encryption(&self) -> SMTPEncryption {
        match self.smtp_secure.as_str() {
            "ssl" => SMTPEncryption::Implicit,
            "tls" => SMTPEncryption::StartTls,
            _ if self.smtp_auto_tls => SMTPEncryption::Opportunistic,
            _ => SMTPEncryption::None,
        }
    }

    fn has_credentials(&self) -> bool {
        !php_empty_str(&self.username) && !php_empty_str(&self.password)
    }

    fn process_email(&self, message: &Email) -> Result<ResponseData, MessagingError> {
        let mut response = Response::new(TYPE);
        let recipients = envelope_recipients(message);

        let transport = match self.connect() {
            Ok(t) => t,
            Err(error) => {
                for email in &recipients {
                    response.add_result(email, error.to_string());
                }
                return Ok(response.to_array());
            }
        };

        let headers = if self.x_mailer.is_empty() {
            Vec::new()
        } else {
            vec![("X-Mailer", self.x_mailer.as_str())]
        };

        let size = Mime::size(message)?;
        if size > 25 * 1024 * 1024 {
            return Err(MessagingError::message(
                "Attachments size exceeds the maximum allowed size of 25MB",
            ));
        }

        let mime = Mime::message(
            message,
            message.get_to(),
            message.get_cc().unwrap_or(&[]),
            message.get_bcc().unwrap_or(&[]),
            &headers,
        );

        let envelope = build_envelope(message)?;
        match transport.send_raw(&envelope, mime.as_bytes()) {
            Ok(_) => {
                response.set_delivered_to(recipients.len() as i64);
                for email in &recipients {
                    response.add_result(email, "");
                }
            }
            Err(error) => {
                for email in &recipients {
                    response.add_result(email, error.to_string());
                }
            }
        }

        if !self.keep_alive {
            self.disconnect();
        }

        Ok(response.to_array())
    }

    fn connect(&self) -> Result<SmtpTransport, MessagingError> {
        if self.keep_alive {
            if let Some(existing) = self.client.lock().clone() {
                return Ok(existing);
            }
        }

        let timeout = Duration::from_secs(self.timeout);
        let mut failures = Vec::new();
        for (host, port, encryption) in self.hosts() {
            if let Err(error) = probe_host(&host, port, timeout) {
                failures.push(format!("{host}:{port} ({error})"));
                continue;
            }
            match build_transport(
                &host,
                port,
                encryption,
                self,
                timeout,
                Duration::from_secs(self.timelimit),
            ) {
                Ok(transport) => {
                    if self.keep_alive {
                        *self.client.lock() = Some(transport.clone());
                    }
                    return Ok(transport);
                }
                Err(error) => {
                    failures.push(format!("{host}:{port} ({error})"));
                }
            }
        }
        Err(MessagingError::message(format!(
            "No SMTP host answered: {}",
            failures.join("; ")
        )))
    }
}

impl Adapter for SMTP {
    fn get_name(&self) -> &'static str {
        "SMTP"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::Email
    }
    fn get_max_messages_per_request(&self) -> usize {
        1000
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(
            self.process_email(expect_email(message)?)?,
        ))
    }
}

fn parse_host_port(entry: &str, default_port: u16) -> (String, u16) {
    // PHP: `/^(\[[^\]]+\]|[^:]+):(\d+)$/`
    if let Some(caps) = split_host_port(entry) {
        return caps;
    }
    (entry.to_string(), default_port)
}

fn split_host_port(entry: &str) -> Option<(String, u16)> {
    if let Some(rest) = entry.strip_prefix('[') {
        let close = rest.find(']')?;
        let host = format!("[{}]", &rest[..close]);
        let after = &rest[close + 1..];
        if let Some(port_str) = after.strip_prefix(':') {
            let port: u16 = port_str.parse().ok()?;
            if after[1..].chars().all(|c| c.is_ascii_digit()) {
                return Some((host, port));
            }
        }
        return None;
    }
    if !entry.contains(':') {
        return None;
    }
    // `[^:]+):(\d+)$` - host with no colon, then :port
    let (host, port) = entry.rsplit_once(':')?;
    if host.contains(':') {
        return None;
    }
    let port: u16 = port.parse().ok()?;
    Some((host.to_string(), port))
}

fn probe_host(host: &str, port: u16, timeout: Duration) -> Result<(), String> {
    let target = format!("{host}:{port}");
    let addrs: Vec<SocketAddr> = target
        .to_socket_addrs()
        .map_err(|e| e.to_string())?
        .collect();
    if addrs.is_empty() {
        return Err("name resolution failed".into());
    }
    TcpStream::connect_timeout(&addrs[0], timeout).map_err(|e| e.to_string())?;
    Ok(())
}

fn build_transport(
    host: &str,
    port: u16,
    encryption: SMTPEncryption,
    smtp: &SMTP,
    timeout: Duration,
    read_timeout: Duration,
) -> Result<SmtpTransport, MessagingError> {
    let host_clean = host.trim_matches(['[', ']']);
    let mut builder = match encryption {
        SMTPEncryption::None => SmtpTransport::builder_dangerous(host_clean).port(port),
        SMTPEncryption::Implicit => {
            let tls = TlsParameters::new(host_clean.to_string())
                .map_err(|e| MessagingError::message(e.to_string()))?;
            SmtpTransport::relay(host_clean)
                .map_err(|e| MessagingError::message(e.to_string()))?
                .port(port)
                .tls(Tls::Wrapper(tls))
        }
        SMTPEncryption::StartTls => {
            let tls = TlsParameters::new(host_clean.to_string())
                .map_err(|e| MessagingError::message(e.to_string()))?;
            SmtpTransport::relay(host_clean)
                .map_err(|e| MessagingError::message(e.to_string()))?
                .port(port)
                .tls(Tls::Required(tls))
        }
        SMTPEncryption::Opportunistic => {
            let tls = TlsParameters::new(host_clean.to_string())
                .map_err(|e| MessagingError::message(e.to_string()))?;
            SmtpTransport::builder_dangerous(host_clean)
                .port(port)
                .tls(Tls::Opportunistic(tls))
        }
    };
    let timeout = timeout.max(read_timeout);
    builder = builder.timeout(Some(timeout));
    if smtp.has_credentials() {
        builder = builder
            .credentials(Credentials::new(
                smtp.username.clone(),
                smtp.password.clone(),
            ))
            .authentication(vec![Mechanism::Plain, Mechanism::Login]);
    }
    Ok(builder.build())
}

fn envelope_recipients(message: &Email) -> Vec<String> {
    let mut seen = Vec::new();
    for list in [
        message.get_to(),
        message.get_cc().unwrap_or(&[]),
        message.get_bcc().unwrap_or(&[]),
    ] {
        for r in list {
            if !seen.contains(&r.email) {
                seen.push(r.email.clone());
            }
        }
    }
    seen
}

fn build_envelope(message: &Email) -> Result<Envelope, MessagingError> {
    let from: Address = message
        .get_from_email()
        .parse()
        .map_err(|e: lettre::address::AddressError| MessagingError::message(e.to_string()))?;
    let to: Vec<Address> = envelope_recipients(message)
        .iter()
        .filter_map(|e| e.parse().ok())
        .collect();
    Envelope::new(Some(from), to).map_err(|e| MessagingError::message(e.to_string()))
}

File: src/Utopia/Messaging/Adapter/Email/SMTP.php
<?php

namespace Utopia\Messaging\Adapter\Email;

use Utopia\Messaging\Adapter\Email as EmailAdapter;
use Utopia\Messaging\Messages\Email as EmailMessage;
use Utopia\Messaging\Response;
use Utopia\SMTP\Auth\Login;
use Utopia\SMTP\Auth\Plain;
use Utopia\SMTP\Client;
use Utopia\SMTP\Encryption;
use Utopia\SMTP\Exception\SmtpException;
use Utopia\SMTP\Exception\TransactionException;
use Utopia\SMTP\Message as SmtpMessage;
use Utopia\SMTP\Timeouts;
use Utopia\SMTP\Transport\Native;

class SMTP extends EmailAdapter
{
    protected const NAME = 'SMTP';

    private ?Client $client = null;

    /**
     * @param string $host SMTP hosts. Either a single hostname or multiple semicolon-delimited hostnames. A port may follow a hostname after a colon (e.g. "smtp1.example.com:25;smtp2.example.com"), and an address literal is bracketed to keep its own colons apart from it (e.g. "[::1]:587"). An encryption prefix may lead each entry (e.g. "tls://smtp1.example.com:587;ssl://smtp2.example.com:465"). Hosts are tried in order.
     * @param int $port The default SMTP server port.
     * @param string $username Authentication username.
     * @param string $password Authentication password.
     * @param string $smtpSecure SMTP Secure prefix. Can be '', 'ssl' or 'tls'
     * @param bool $smtpAutoTLS Enable/disable SMTP AutoTLS feature. Defaults to false.
     * @param string $xMailer The value to use for the X-Mailer header.
     * @param int $timeout SMTP timeout in seconds.
     * @param bool $keepAlive Whether to reuse the SMTP connection across process() calls.
     * @param int $timelimit SMTP command timelimit in seconds.
     * @param int $pingThreshold Seconds a kept session may sit idle before it is probed ahead of the next message. Keep it under the server's idle timeout: a session the server closed inside the threshold fails its send without a retry. 0 probes before every reuse.
     * @param int $restartThreshold Messages a kept session carries before it is replaced. 0 disables.
     */
    public function __construct(
        private readonly string $host,
        private readonly int $port = 25,
        private readonly string $username = '',
        private readonly string $password = '',
        private readonly string $smtpSecure = '',
        private readonly bool $smtpAutoTLS = false,
        private readonly string $xMailer = '',
        private readonly int $timeout = 30,
        private readonly bool $keepAlive = false,
        private readonly int $timelimit = 30,
        private readonly int $pingThreshold = 30,
        private readonly int $restartThreshold = 100,
    ) {
        parent::__construct();
        if (!\in_array($this->smtpSecure, ['', 'ssl', 'tls'])) {
            throw new \InvalidArgumentException('Invalid SMTP secure prefix. Must be "", "ssl" or "tls"');
        }
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
     * {@inheritdoc}
     */
    protected function process(EmailMessage $message): array
    {
        $response = new Response($this->getType());
        $recipients = $this->recipients($message);

        try {
            $client = $this->keepAlive ? $this->client() : $this->dial();
        } catch (SmtpException $exception) {
            foreach ($recipients as $email) {
                $response->addResult($email, $exception->getMessage());
            }

            return $response->toArray();
        }

        $keep = $this->keepAlive;

        try {
            $result = $client->send($this->build($message));

            // A server may refuse some recipients and accept others, and the
            // message still reaches the rest. Each address gets its own answer
            // rather than every address getting the same one.
            $response->setDeliveredTo(\count($result->accepted));

            foreach ($result->accepted as $email) {
                $response->addResult($email);
            }

            foreach ($result->rejected as $email => $reply) {
                $response->addResult($email, (string) $reply);
            }
        } catch (TransactionException $exception) {
            foreach ($recipients as $email) {
                $response->addResult($email, (string) $exception->reply);
            }

            // A 421 during RCPT ends the session with the transaction.
            $keep = $keep && is_finite($client->idle());
        } catch (SmtpException $exception) {
            foreach ($recipients as $email) {
                $response->addResult($email, $exception->getMessage());
            }

            $keep = false;
        } finally {
            if (!$keep) {
                $this->drop($client);
            }
        }

        return $response->toArray();
    }

    /**
     * Teardown only: closing under a send in flight desynchronises it.
     */
    public function disconnect(): void
    {
        $this->client?->close();
        $this->client = null;
    }

    /**
     * The kept session, or a fresh one when it is spent or the server closed it.
     */
    private function client(): Client
    {
        if ($this->client instanceof Client && $this->reusable($this->client)) {
            return $this->client;
        }

        $this->client?->close();
        $this->client = null;

        return $this->client = $this->dial();
    }

    private function drop(Client $client): void
    {
        if ($this->client === $client) {
            $this->client = null;
        }

        $client->close();
    }

    /**
     * The first host that answers, tried in the order they were given.
     */
    private function dial(): Client
    {
        $timeouts = new Timeouts(
            connect: (float) $this->timeout,
            read: (float) $this->timelimit,
            write: (float) $this->timelimit,
        );

        $failures = [];

        foreach ($this->hosts() as [$host, $port, $encryption]) {
            $client = new Client(
                new Native($host, $port),
                gethostname() ?: 'localhost',
                $this->authenticators(),
                $encryption,
                $timeouts,
            );

            try {
                // Nothing is dialled until a session is needed, so asking what
                // the server offers is what settles whether this host answers.
                $client->capabilities();
            } catch (SmtpException $exception) {
                $failures[] = "{$host}:{$port} ({$exception->getMessage()})";

                continue;
            }

            return $client;
        }

        throw new \Utopia\SMTP\Exception\ConnectionException(
            'No SMTP host answered: ' . implode('; ', $failures),
        );
    }

    private function reusable(Client $client): bool
    {
        if ($this->restartThreshold > 0 && $client->transactions >= $this->restartThreshold) {
            return false;
        }

        if ($client->idle() <= $this->pingThreshold) {
            return true;
        }

        return $client->ping();
    }

    /**
     * The host string, which may name several servers with their own port and
     * encryption, as a list of somewhere to try.
     *
     * @return list<array{string, int, Encryption}>
     */
    private function hosts(): array
    {
        $hosts = [];

        foreach (explode(';', $this->host) as $entry) {
            $entry = trim($entry);

            if ($entry === '') {
                continue;
            }

            $encryption = $this->encryption();

            if (preg_match('#^(ssl|tls)://#i', $entry, $matches) === 1) {
                $encryption = strtolower($matches[1]) === 'ssl' ? Encryption::Implicit : Encryption::StartTls;
                $entry = substr($entry, \strlen($matches[0]));
            }

            $port = $this->port;

            // An address literal is mostly colons, so only a bracketed one or a
            // name carrying none may be followed by a port. PHPMailer's own
            // pattern reads "::1" as the host ":" on port 1.
            if (preg_match('/^(\[[^\]]+\]|[^:]+):(\d+)$/', $entry, $matches) === 1) {
                $entry = $matches[1];
                $port = (int) $matches[2];
            }

            $hosts[] = [$entry, $port, $encryption];
        }

        return $hosts === [] ? [[$this->host, $this->port, $this->encryption()]] : $hosts;
    }

    /**
     * The prefix says what to do, and without one the AutoTLS flag decides
     * whether an offered upgrade is taken.
     */
    private function encryption(): Encryption
    {
        return match ($this->smtpSecure) {
            'ssl' => Encryption::Implicit,
            'tls' => Encryption::StartTls,
            default => $this->smtpAutoTLS ? Encryption::Opportunistic : Encryption::None,
        };
    }

    /**
     * @return list<Plain|Login>
     */
    private function authenticators(): array
    {
        // The adapter this replaces treated a literal "0" as no credential,
        // an artefact of empty() that some configuration still relies on.
        if (\in_array($this->username, ['', '0'], true) || \in_array($this->password, ['', '0'], true)) {
            return [];
        }

        return [new Plain($this->username, $this->password), new Login($this->username, $this->password)];
    }

    /**
     * Every address the envelope will carry, which is what a per-recipient
     * result is keyed by.
     *
     * @return list<string>
     */
    private function recipients(EmailMessage $message): array
    {
        $recipients = [];

        foreach ([...$message->getTo(), ...($message->getCC() ?? []), ...($message->getBCC() ?? [])] as $recipient) {
            $recipients[$recipient['email']] = true;
        }

        return array_keys($recipients);
    }

    private function build(EmailMessage $message): SmtpMessage
    {
        $headers = [];

        if ($this->xMailer !== '') {
            $headers['X-Mailer'] = $this->xMailer;
        }

        if (Mime::size($message) > self::MAX_ATTACHMENT_BYTES) {
            throw new \Exception('Attachments size exceeds the maximum allowed size of 25MB');
        }

        return Mime::message(
            $message,
            $message->getTo(),
            $message->getCC() ?? [],
            $message->getBCC() ?? [],
            $headers,
        );
    }
}
